src/main/java/com/google/codeu/servlets/UserFormServlet.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.codeu.servlets;
+
+import java.io.IOException;
+
+
+import javax.servlet.ServletConfig;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.codeu.data.Datastore;
+import com.google.codeu.data.Tea;
+
+
+
+
+public class UserFormServlet extends HttpServlet {
+  
+  private Datastore datastore;
+
+  @Override
+  public void init(ServletConfig config) throws ServletException {
+    super.init(config);
+    datastore = createDatastore();
+  }
+
+  /*
+   * Instantiates a connection to a datastore. In tests, override this with a
+   * method that returns a fake/mock datastore object.
+   */
+  protected Datastore createDatastore() {
+    return new Datastore();
+  }
+
+@Override
+public void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+
+  UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/index.html");
```

Comment:
This is a soft failure. It is better to do a hard failure by emitting a `401` status code (which means "unauthenticated").

To make this a good user experience with a hard failure, the `POST` should be done in JavaScript (where the JavaScript code can handle the failure), rather than having this URL be hit directly by the browser as part of the form's action. However, although it means a little more client-side work for the web case, it is still worth using a hard failure rather than a soft failure here, because it then makes client-side detection of the error much more reliable for other kinds of clients (e.g. if you were to create a mobile application or a script that hit this endpoint, redirects are typically followed automatically and are easily miscategorized as "success", while setting an appropriate error status code will make it easy for code that fetches the page programatically to realize that something has gone wrong).